### PR TITLE
[RFC] [Needs testing and sweep] Remove C++ guards

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94973
* #94970

cc @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/guards.py [hunk 1]
```diff
@@ -410,36 +395,31 @@ def TENSOR_MATCH(self, guard: Guard):
...
-                self._produce_guard_code(guard, code)
-            else:
-                self.tensor_check_names.append(tensor_name)
-                self.tensor_check_examples.append(value)
+            terms = ["dtype", "device.index", "requires_grad", "ndimension()"]
+            if not config.dynamic_shapes:
+                terms.extend(["size()", "stride()"])
+
+            code = []
+            # TODO(voz): Ask ed how to turn a keyset like DispatchKeySet(CUDA, ADInplaceOrView, AutogradCUDA, AutocastCUDA) into a
```
Comment (voznesenskym) [inline torch/_dynamo/guards.py:414]: @ezyang help

Comment (ezyang) [reply torch/_dynamo/guards.py:414]: There's a crappy binding at `_dispatch_keys` in `torch/csrc/utils/python_dispatch.cpp` but it is probably not very fast.

**Comment by ezyang** [comment]
So, @jansel in code comments says he benchmarked this and the Python guards were too slow. I tend to trust jansel's benchmarking numbers. So @voznesenskym, you're going to have to explain what is different; why can we remove this now, whereas jansel had to *go out of his way* to implement C++ guards in the first place.

**Review by jansel — CHANGES_REQUESTED** [review]
This doesn't seem ready, it needs:
1) Benchmark results in the description
2) Fix all the test errors

**Comment by voznesenskym** [comment]
Initial sweep - 98% - same speed? https://docs.google.com/spreadsheets/d/1HneGfS07on1xhYO4_Kv47zVxNYEoJ7mz5niEBLC9FGc/edit?usp=sharing

**Comment by voznesenskym** [comment]
Need to find some real world models to try this on

**Comment by ezyang** [comment]
Aren't you measuring the wrong thing. frame_time measures 

```
@dynamo_timed(phase_name="entire_frame_compile")
def _compile(
    code: types.CodeType,
    globals: Dict[str, object],
    locals: Dict[str, object],
    builtins: Dict[str, object],
    compiler_fn: CompilerFn,
    one_graph: bool,
    export: bool,
    hooks: Hooks,
    frame: Optional[types.FrameType] = None,
) -> Optional[GuardedCode]:
```

but you changed the guard execution logic. You need to measure how fast it is AFTER warmup. In fact, the E2E performance benchmark should work for your use case.
